lib/controllers/adm/services.ts

```diff
@@ -0,0 +1,228 @@
+    const errorOrMaybeService = await serviceModel.findOneByServiceId(
+      serviceId
+    );
+    if (errorOrMaybeService.isLeft) {
+      return ResponseErrorQuery(
+        "Error trying to retrieve existing service",
+        errorOrMaybeService.left
+      );
+    }
+
+    const maybeService = errorOrMaybeService.right;
+    if (maybeService.isEmpty) {
+      return ResponseErrorNotFound(
+        "Error",
+        "Could not find a service with the provided serviceId"
+      );
+    }
+
+    const existingService = maybeService.get;
+
+    const errorOrMaybeUpdatedService = await serviceModel.update(
+      existingService.id,
+      existingService.serviceId,
+      currentService => {
+        const updatedService = {
+          ...currentService,
+          ...serviceModelPayload,
+          serviceId
+        };
+        return updatedService;
+      }
+    );
+    if (errorOrMaybeUpdatedService.isLeft) {
+      return ResponseErrorQuery(
+        "Error while updating the existing service",
+        errorOrMaybeUpdatedService.left
+      );
+    }
+
+    const maybeUpdatedService = errorOrMaybeUpdatedService.right;
+    if (maybeUpdatedService.isEmpty) {
+      return ResponseErrorValidation(
+        "Error while updating the existing service",
+        "Cannot find the updated service"
+      );
+    }
+
+    return ResponseSuccessJson(maybeUpdatedService.get);
+  };
+}
+
+export function CreateServiceHandler(
+  serviceModel: ServiceModel,
+  generateObjectId: ObjectIdGenerator
+): ICreateServiceHandler {
+  return async (_, serviceModelPayload) => {
+    if (serviceModelPayload.serviceId) {
+      return ResponseErrorValidation(
+        "Error validating payload",
+        "service_id must be empty or undefined"
+      );
+    }
+    const serviceId = generateObjectId();
+    const errorOrService = await serviceModel.create(
+      { ...serviceModelPayload, serviceId },
+      serviceId
+    );
+    if (errorOrService.isRight) {
+      return ResponseSuccessJson(errorOrService.right);
+    } else {
+      return ResponseErrorQuery("Error", errorOrService.left);
+    }
+  };
+}
+
+/**
+ * Wraps a CreateService handler inside an Express request handler.
+ */
+export function CreateService(
+  serviceModel: ServiceModel
+): express.RequestHandler {
+  const handler = CreateServiceHandler(serviceModel, ulidGenerator);
+  const middlewaresWrap = withRequestMiddlewares(
+    AzureApiAuthMiddleware(new Set([UserGroup.ApiServiceWrite])),
+    ServicePayloadMiddleware
+  );
+  return wrapRequestHandler(middlewaresWrap(handler));
+}
+
+/**
+ * Wraps an UpdateService handler inside an Express request handler.
+ */
+export function UpdateService(
+  serviceModel: ServiceModel
+): express.RequestHandler {
+  const handler = UpdateServiceHandler(serviceModel);
+  const middlewaresWrap = withRequestMiddlewares(
+    AzureApiAuthMiddleware(new Set([UserGroup.ApiServiceWrite])),
+    RequiredParamMiddleware(params => {
```

Comment:
for clarity, pls instantiate this middleware outside `UpdateService`, i.e. `const requiredServiceIdMiddleware = ...`

Reply:
ok